Expand parallel property section in the README (#9)
stevana/quickcheck-state-machine: README.md

race conditions -- which normally can be tricky to test for. It works as
follows:

  1. generate a list of actions and split it in two (or more) parts:

       - a first part that will be run sequentially, called the _prefix_ (think
         of this as an initialisation bit that setups up some state);
       - a second part (the _suffix_) that will be split in sublists which will
         be run in parallel (see `parallelSafe` to understand how it determines
         that a sequence of commands can be run in parallel). More than one
         suffix can be generated, i.e. this second step can be done multiple
         times with the part of the generated list that doesn't belong to the
         prefix.

>   2. execute the prefix sequentially as in the section above (checking pre and

Reviewer: ```suggestion
  2. execute the prefix sequentially as in the section above (checking pre- and
```

     and gather the trace (or history) of invocations and responses of each
     action;

```
                            ┌── no checks aside from ensuring no exception was thrown
                            │
                  ╭─────────┴──────────╮

                  ┌─ [C] ──┐  ┌ [F, G] ┐   ◀─╮
Commands: [A, B] ─┤        ├──┤        │     ├─ executed `concurrently`
                  └ [D, E] ┘  └ [H, I] ┘   ◀─╯

          ╰─┬──╯       ▲          ▲
            │          ╰─────┬────╯
            │                └── groups are not run in parallel
            │                    i.e [C, D, E] will run (and
            │                    finish) before F or H are
            │                    started
            │
            └── pre/postconditions and invariant checked
                executed sequentially
```

  4. if something goes wrong when executing the commands, shrink the generated
     commands and present a minimal counterexample;

  5. otherwise, try to find a possible sequential interleaving of action
     invocations and responses that respects the post-conditions. For each
     interleaving, this is done by advancing the `Concrete` model (starting at

>      `initalModel`) through the sequence of pairs of invocations to `command

Reviewer: ```suggestion
     `initialModel`) through the sequence of pairs of invocations to `command
```

>      Concrete` and the returned `response Concrete` emitted at step 3, check the

Reviewer: ```suggestion
     Concrete` and the returned `response Concrete` emitted at step 3, and checking the
```

>      postcondition at each pair.

Reviewer: ```suggestion
     post-condition for each pair.
```

  6. if no possible sequential interleaving was found, then shrink the generated
     commands and present a minimal counterexample.

> The last step basically tries to find

Reviewer: ```suggestion
The last two steps basically tries to find
```

a [linearisation](https://en.wikipedia.org/wiki/Linearizability) of calls that
could have happend on a single thread.

Notice that step 5 above introduces a subtlety in the post-condition checks and
transitions for the `model Concrete`. Despite the system under test running in a

> concurrent way, the model can still be designed to work sequentially as *it will

Reviewer: Yes, this is a key point of linearisability checking. It's also an assumption: the model most be correct with respect to sequential execution before it's used in concurrent testing.

Author: This is an important point, in my last commit I added this to the paragraph.

not be run in parallel*, it will only be advanced in a sequential way when
evaluating possible interleavings.

As we cannot control the actual scheduling of the tasks, the sequence of
commands (already fixed in a concrete prefix and a concrete list of suffixes) is
actually executed several times by default, i.e. steps 2 to 6 will be executed
multiple times for the same test case expecting that the scheduling of events

> varies between runs.

Reviewer: Note that one can further introduce entropy here by introducing random sleeps in the semantic function.

Author: I added this to the end of the paragraph, I think it is an important subtlety worth mentioning

     checking the post-condition for each pair.

  6. if no possible sequential interleaving was found, then shrink the generated
     commands and present a minimal counterexample.

The last two steps basically try to find
a [linearisation](https://en.wikipedia.org/wiki/Linearizability) of calls that
could have happend on a single thread.

Notice that step 5 above introduces a subtlety in the post-condition checks and
transitions for the `model Concrete`. Despite the system under test running in a
concurrent way, the model can still be designed to work sequentially as *it will
not be run in parallel*, it will only be advanced in a sequential way when
evaluating possible interleavings. This particularly means that the model must
be correct with respect to sequential execution before used in parallel testing.

As we cannot control the actual scheduling of the tasks, each sequence of
commands (already fixed in a concrete prefix and a concrete list of suffixes) is
actually executed several times by default, i.e. steps 2 to 6 will be executed
multiple times for the same test case expecting that the scheduling of events

> varies between runs. One can increase entropy by introducing random `sleep`s in

Reviewer: ```suggestion
varies between runs. One can further increase entropy by introducing random `threadDelay`s in
```
